src/irmin/tree_intf.ml

```diff
@@ -102,22 +120,24 @@ module type S = sig
 
   (** {1 Manipulating Subtrees} *)
 
-  (** Operations on {{!node} tree nodes}. *)
+  (** Operations on {{!node} lazy tree nodes}. *)
   module Node : sig
     type t = node
-    (** The type of tree nodes. *)
+    (** The type of lazy tree nodes. *)
 
     val t : t Type.t
     (** The value type of tree nodes. *)
 
-    val hash : t -> hash
+    val to_hash : t -> hash
```

Comment:
We should make sure it's the same as `Irmin.Type.hash t`

Reply:
It's not: the value-type describes the lazy view (so hash will vary according to `Map | Hash | Value`) whereas the `hash` function gives the hash of the underlying node in the store (if it exists). This is why I wasn't sure whether `to_hash` was a better name for this function.

Reply:
so maybe `t` should be changed to use `hash` as custom hash function.

Reply:
Hmm, I'm not sure about that. `hash` isn't actually hashing `t` values: it's hashing the thing that the `t` _points to_. That would result in situations where `Type.(hash t) n1 = Type.(hash t) n2` but `Type.(equal t) n1 n2 = false`, which _looks_ like a collision (but isn't actually).

How about overriding the generic `hash`  to a function raising an exception?

Reply:
That doesn't shock me: if two nodes are different but they have the same hash, that just mean that they represent the same object but have slightly different runtime representation. We could also imagine in the future that we could hashcons all the objects with the same hash ; in that case the equality between objets is not so important. I agree we should document this in order to make this less confusing, but I find the invariant "hash(tree) = hash(object pointed by the tree)" a good one.

An other option is just to not expose `val t : t Irmin.Type.t` at all...

Reply:
It does confuse me to have certain generic operations being delegated to the represented object when others refer to the handle. I'm happy to just not expose this :slightly_smiling_face:

Reply:
I agree this is confusing :-) But it's either this or we should stay consistent but make sure the "correct" (e.g. the new one) `hash` function in `val t : t Irmin.Type.t`.